File: truck_tour.js
function truckTour(petrolpumps) {
    // counter indicating current station we're attempting
    let pump = 0;
    //init gas counter
    let tank = 0;
    // loop through every element in the pumps array
    for (let i = 0; i < petrolpumps.length; i++) {
        // we want to start at the first pump and go in order
        // to find the station with the smallest index

        // increment counter to track how much gas we have
        // if gas counter >= distance,
        //then we move on to the next station
        tank += (petrolpumps[i][0] - petrolpumps[i][1])

        if (tank < 0) {
            // else
            // reset and try next station
            // update the counter indicating  which current
            // station we're attempting to start from
            pump = i + 1
            // reset the gas counter to 0
            tank = 0
        }
    }
    // at this point, traverse over the pumps array again
    // starting at the beginning up until where we stopped
    // checking that the solution still works
    for (let i = 0; i < pump; i++) {
        tank += (petrolpumps[i][0] - petrolpumps[i][1])
        if (tank < 0) {
            // else
            // reset and try next station
            // update the counter indicating which current
            // station we're attempting to start from
            pump = i + 1
            // reset the gas counter to 0
            tank = 0
        }
    }
    return pump
}


// function truckTour(petrolpumps) {
//     // queue to keep track of pumps
//     // if we realize pump is not valid starting point
//     // wee need to add it back to the queue so that it
//     // gets considered later on when we're trying
//     // a different starting pump

//     const route = []
//     // enqueue every pump in the array
//     for (let i = 0; i < petrolpumps.length; i++) {
//         route.push(petrolpumps[i])
//     }
//     // starting pump number
//     let pump = 0
//     let tank = 0
//     // what's our ending criteria?
//     // let's keep track of the number of pumps we've passed
//     let passed = 0

//     // loop until passed  === number of pumps we have
//     while (passed < petrolpumps.length) {
//         // get first pump from the route
//         const current = route.shift()
//         // add the amount of gas to our tank
//         tank += current[0]
//         // check if we can make it to the next station
//         if (tank - current[1] >= 0) {
//             // then we can successfully count
//             // this station as "passed"
//             passed++
//             // decrement gas counter by the amount used to get
//             // to the next station
//             tank -= current[1]
//         } else {
//             // reset and consider the next pump
//             pump += passed + 1
//             tank = 0
//         }
//         // add the current pump back onto our route
//         route.push(current)
//     }
//     return pump
// }

File: truck_tour.py
import queue

def truckTour(petrolpumps):
    # init queue 
    route = queue.Queue()

    # add all the pumps to the queue 
    for p in petrolpumps:
        route.put(p) 
    # var to hold the pump we end up starting at 
    start = 0
    # var to keep track of the number of pumps we've traversed 
    traversed = 0
    # amount of gas we have 
    gas = 0

    # loop so long as we haven't traversed every single pump 
    while traversed < len(petrolpumps):
        # get the next pump along the route 
        pump = route.get() 
        # add the amount of gas it has available 
        gas += pump[0]
        # check if our gas tank has enough to get us 
        # to the next pump 
        if gas >= pump[1]:
            # increment traversed counter 
            traversed += 1
            # decrement by the amount of gas it takes to get there 
            gas -= pump[1] 
        # if it doesn't, reset our gas tank, the number of pumps
        # we've traversed, and move on to consider the next pump
        else:
            # otherwise, starting at the pump we started at 
            # isn't a viable option, so let's consider the 
            # pump after the one we're currently on 
            start += traversed + 1
            # reset traversed counter 
            traversed = 0
            # reset gas counter 
            gas = 0
        # add the pump to the back of the route 
        route.put(pump)
    # at this point, we've found the starting pump that will
    # allow us to traverse all of the pumps; return it
    return start